File: hangman-beta.py
#imports
import hashlib, random

# configurations
hashprependConfig = "no password for u "
wonConfig = 0
triesConfig = 5
hintsConfig = 1
loginFailuresConfig = 3
menuMsgConfig = "\nWelcome to The HANGHANGMAN\n\nSelect an option\n1. Leaderboard\n2. Log In\n3. Play as Guest\n4. Create an Account\n5. Game Credits\n(Please enter the number): "
catquesConfig = '\nPlease choose a category'
enterNumberPromptConfig = "\n(Please enter the number): "
guessPromptConfig = "\nGuess a letter or the entire word: "
playAgainPromptConfig = "\nPlay again? (y/n): "
hintPromptConfig = "Would you like a hint? (y/n): "
wonMsgConfig = "\n\t--------------------\n\t-- You have won!! --\n\t--------------------"
lostMsgConfig = "\n\t-------------------\n\t-- You have lost --\n\t-------------------"
errorMsg1Config = "Please only enter an alphabet or a guess with the same length as the word"
errorMsg2Config = "Please enter y or n only"
errorMsg3Config = "\nPlease enter one of the above options"
errorMsg4Config = "\nUser has to be at least 4 characters and Password has to be at least 8 characters"
errorMsg5Config = "\nThis username is taken"
endMsgConfig = "END"
noDataConfig = "There is no data"
tryAgnConfig = "Please try again"
creditsConfig = "\n\nCredits\nAuthor: Ao Yu"
createUsernameConfig = "Please enter credentials\nUser: "
createUserpassConfig = "Password: "
confirmUserpassConfig = "Enter Password again: "
leaderboardConfig = "\nLeaderboard:\n"
regex = '/^[A-Za-z]+$/'

# declaration of values
dictionary = []
selecteddictionary = []
arrayOfChosenWord = []
wordLinesArray = []
categories = []
categoriesName = []
pointsarr = []
wordLines = ""
guess = ""
hint = ""
categoryChosen = ""
loginId = 0
usersPoints = 0

# assigning configs to values
won = wonConfig
tries = triesConfig
hints = hintsConfig
catques = catquesConfig
loginFailures = loginFailuresConfig
enterNumberPrompt = enterNumberPromptConfig
guessPrompt = guessPromptConfig
playAgainPrompt = playAgainPromptConfig
hintPrompt = hintPromptConfig

# messages
def wonMsg(): print(wonMsgConfig)
def lostMsg(): print(lostMsgConfig)
def errorMsg1(): print(errorMsg1Config)
def errorMsg2(): print(errorMsg2Config)
def triesLeftMsg(tries): print(f'You have {tries} lives left')
def endMsg(): print(endMsgConfig)
def noData(): raise Exception(noDataConfig)
def tryAgn(): print(tryAgnConfig)




def sortFunction(a,b):
  if a[1] == b[1]:
    return 0
  else:
    return
    # return (a[1] < b[1] ? 1: -1)

def readDict():
  data = open("dictionary.txt", "r")
  if data.read() == "":
    noData()
  else:
    dictionary = data.split('\n')
    return dictionary
    # for i in dictionary.length:
    #   dictionary[i] = 

def menu():
  choice = input(menuMsgConfig)

  if choice == "1":
    leaderboard()
  elif choice == "2":
    login()
  elif choice == "3":
    main()
  elif choice == "4":
    createUser()
  elif choice == "5":
    print(creditsConfig)
    menu()
  else:
    print(errorMsg3Config)
    choice = input(menuMsgConfig)


def hashencrypt(password):
  password = hashprependConfig + password
  password = hashlib.md5(password.encode())
  return password

def createUser():
  data = open("dictionary.txt", "r")
  if data.read() == "":
    noData()
  else:
    createUsername = input(createUsernameConfig)
    createUserpass = input(createUserpassConfig)
    confirmUserpass = input(confirmUserpassConfig)
    loginCreds = data.split('\n')
    for i in loginCreds.length:
      newId = loginCreds[loginCreds.length - 1].id + 1
      if createUsername == loginCreds[i].name:
        print(errorMsg5Config)
        createUser()
    if createUsername.length >= 4 and createUserpass.length >= 8 and createUserpass == confirmUserpass:
      f = open("users.txt", "a")
      f.write(f'\n\{"id":{newId},"name":"{createUsername}","password":"{hashencrypt(createUserpass)}","score":0,"admin":0}')
      f.close()
    else:
      print(errorMsg4Config)

def leaderboard():
  data = open("users.txt", "r")
  if data.read() == "":
    noData()
  else:
    pointslbdata = data.split('\n')
    for i in pointslbdata.length:
      if pointslbdata[i].score != 0:
        pointsarr.append([{pointslbdata[i].name}, pointslbdata[i].score])
    pointsarr.sort()
    # pointsarr = '\n'.join(pointsarr)
    # pointsarr = pointsarr.replace()
    print(pointsarr)

def loginFail():
  loginFailures -= 1
  tryAgn()
  login()

def login():
  loginDetailsUser = input("\nUser: ")
  loginDetailsPassw = input("Password:")
  data = open("users.txt", "r")
  if data.read() == "":
    noData()
  else:
    loginCred = data.split('\n')
    for i in loginCred.length:
      if loginCred[i].name == loginDetailsUser and loginCred[i].password == hashencrypt(loginDetailsPassw):
        loginId = loginCred[i].id
        main()
        return

    loginFail()

def points():
  data = open("users.txt", "r")
  if data.read() == "":
    noData()
  else:
    pointsdata = data.split('\n')
    for i in pointsdata.length:
      if loginId == pointsdata[i].id:
        pointsdata[i].score += 100
    f = open("users.txt", "a")
    f.write(pointsdata)
    f.close()

def reset():
  selecteddictionary = []
  arrayOfChosenWord = []
  wordLinesArray = []
  catques = catquesConfig
  won = wonConfig
  tries = triesConfig
  hints = hintsConfig

def replay():
  replay = ""
  while "y" not in replay and "n" not in replay:
    replay = input(playAgainPrompt).lower()
    if "y" in replay:
      main()
    elif "n" in replay:
      endMsg()
    else:
      errorMsg2

def pickCat():
  readDict()
  for t in dictionary.length:
    categories[t] = dictionary[t].cat
    categoriesName[t] = dictionary[t].catName


def randomWordPicker():
  for i in dictionary.length:
    if dictionary[i].cat == categoryChosen:
      selecteddictionary.push(dictionary[i])
  randomNumber= random.randint(0, selecteddictionary.length-1)
  chosenWord = selecteddictionary[randomNumber].name
  for i in chosenWord.length:
    arrayOfChosenWord[i] = chosenWord.charAt(i).upper()
  for e in arrayOfChosenWord.length:
    if arrayOfChosenWord[e] == " " or arrayOfChosenWord[e] == "-":
      wordLinesArray[e] = arrayOfChosenWord[e]
    else:
      wordLinesArray[e] = "_"
  wordLines = wordLinesArray.join(" ")
  print("\n", wordLines)
  return randomNumber

def getHint(randomNumber):
  hintOfChosenWord = selecteddictionary[randomNumber].hint
  needHint = input(hintPrompt)
  if needHint.includes("y"):
    print(hintOfChosenWord)
  elif needHint.includes("n"):
    return
  else:
    errorMsg2()
    getHint(randomNumber)

def guessWord():
  if arrayOfChosenWord.join("").includes(guess.upper()):
    if guess.length == 0:
      errorMsg1
    elif guess.length >= 1 and guess.split("").length != arrayOfChosenWord.length:
      tries -= 1
      triesLeftMsg(tries)
      errorMsg1()
    elif guess.length == 1:
      for i in arrayOfChosenWord.length:
        if arrayOfChosenWord[i].includes(guess.upper()):
          wordLinesArray[i] = arrayOfChosenWord[i]
          if wordLinesArray.includes("_"):
            print("\n", wordLinesArray.join(" "))
          elif won == 0:
            won = 1
            print("\n", wordLinesArray.join(" "))
            points()
            wonMsg()
            replay()
          else:
            print("error 1")
    else:
      errorMsg1()
  elif regex.test(guess) == True:
    tries -= 1
    triesLeftMsg(tries)
  else:
    errorMsg1()

def main():
  reset()
  pickCat()
  randomNumber = randomWordPicker()
  while tries > 0 and won == 0:
    if tries == 3 and hints > 0:
      getHint(randomNumber)
      hints -= 1
    guessWord()
  if tries == 0:
    lostMsg()
    replay()

menu()

File: hangman-beta.js
//imports
const fs = require('fs')
const readline = require('readline-sync')
const crypto = require('crypto');
const config = require('./config.json')


const regex = /^[A-Za-z]+$/

// declaration of values
var dictionary = new Array
var selecteddictionary = new Array
var arrayOfChosenWord = new Array
var wordLinesArray = new Array
var categories = new Array
var categoriesName = new Array
var wordLines = new String
var guess = new String
var hint = new String
var categoryChosen = new String

// assigning configs to values
var won = config.wonConfig
var tries = config.triesConfig
var hints = config.hintsConfig
var catques = config.catquesConfig
var loginFailures = config.loginFailuresConfig
var enterNumberPrompt = config.enterNumberPromptConfig
var guessPrompt = config.guessPromptConfig
var playAgainPrompt = config.playAgainPromptConfig
var hintPrompt = config.hintPromptConfig

// messages
function wonMsg() {console.log(config.wonMsgConfig)}
function lostMsg() {console.log(config.lostMsgConfig)}
function errorMsg1() {console.log(config.errorMsg1Config)}
function errorMsg2() {console.log(config.errorMsg2Config)}
function triesLeftMsg(tries) {console.log(`You have ${tries} lives left`)}
function endMsg() {console.log(config.endMsgConfig)}
function noData() {throw new Error(config.noDataConfig)}
function tryAgn() {console.log(config.tryAgnConfig)}

// reading dictionary from file
function readDict() {
    var data = fs.readFileSync("dictionary.txt", 'utf8');
    if (data == "") {
        noData()
    }
    else {
        dictionary = data.split('\n')
        for (i = 0; i < dictionary.length; i++) {
            dictionary[i] = JSON.parse(dictionary[i])
        }
    }
}

// main menu
function menu() {
    var choice = readline.question(config.menuMsgConfig)
    switch (choice) {
    case "1":
        login();
        break;
    case "2":
        main();
        break;
    case "3":
        createUser();
        break;
    case "4":
        console.log(config.creditsConfig)
        menu()
        break;
    default:
        console.log(config.errorMsg3Config)
        var choice = readline.question(config.menuMsgConfig)
    }
}

// hashing password
function hashencrypt(password) {
    password = config.hashprependConfig + password
    password = crypto.createHash('md5').update(password).digest('hex')
    return password
}

// create user
function createUser() {
    var data = fs.readFileSync("users.txt", 'utf8');
    if (data == "") {
        noData()
    }
    else {
        var createUsername = readline.question(config.createUsernameConfig)
        var createUserpass = readline.question(config.createUserpassConfig)
        var confirmUserpass = readline.question(config.confirmUserpassConfig)
        loginCreds = data.split('\n')
        for (i = 0; i < loginCreds.length; i++) {
            loginCreds[i] = JSON.parse(loginCreds[i])
            var newId = loginCreds[loginCreds.length - 1].id + 1
            if (createUsername == loginCreds[i].name) {
                console.log(config.errorMsg5Config)
                createUser()
            }
        }
        if(createUsername.length >= 4 && createUserpass.length >= 8 && confirmUserpass === createUserpass){
                var data = fs.appendFileSync("users.txt", `\n{"id": ${newId}, "name": "${createUsername}", "password": "${hashencrypt(createUserpass)}", "score": 0, "admin": 0}`);
        }
        else {
            console.log(config.errorMsg4Config)
        }
    }
}

// login failures function
function loginFail() {
    loginFailures -= 1
    tryAgn()
    login()
}

// login
function login() {
    var loginDetailsUser = readline.question("\nUser: ")
    var loginDetailsPassw = readline.question("Password: ")
    var data = fs.readFileSync("users.txt", 'utf8');
    if (data == "") {
        noData()
    }
    else {
        loginCred = data.split('\n')
        for (i = 0; i < loginCred.length; i++) {
            loginCred[i] = JSON.parse(loginCred[i])
            if (loginCred[i].name == loginDetailsUser && loginCred[i].password == hashencrypt(loginDetailsPassw)) {
                main()
            }
        }
        loginFail()
    }
}

// resetting values for next game
function reset() {
    selecteddictionary = []
    arrayOfChosenWord = []
    wordLinesArray = []
    catques = config.catquesConfig
    won = config.wonConfig
    tries = config.triesConfig
    hints = config.hintsConfig
}

// for replaying of game
function replay() {
    var replay = ""
    while (!replay.includes("y") && !replay.includes("n")) {
        replay = readline.question(playAgainPrompt).toLowerCase()
        if (replay.includes("y")) {
            main()
        }
        else if (replay.includes("n")) {
            endMsg()
        }
        else {
            errorMsg2()
        }
    }
}

// picking a category to play
function pickCat() {
    readDict()
    for (var t = 0; t < dictionary.length; t++) {
        categories[t] = dictionary[t].cat
        categoriesName[t] = dictionary[t].catName
    }
    var categorieTypes = categories.filter((x, i, a) => a.indexOf(x) === i)
    var categoriesNameTypes = categoriesName.filter((x, i, a) => a.indexOf(x) === i)

    for (var i = 0; i < categorieTypes.length; i ++) {
        var strr = `\n${categorieTypes[i]}: ${categoriesNameTypes[i]}`
        catques = catques.concat(strr)
    }
    categoryChosen = readline.question(catques.concat(enterNumberPrompt))
    return categoryChosen
}

// random word generator from dictionary
function randomWordPicker() {
    for (i = 0; i < dictionary.length; i++) {
        if (dictionary[i].cat == categoryChosen) {
            selecteddictionary.push(dictionary[i])
        }
    }
    var randomNumber = Math.floor((Math.random() * selecteddictionary.length));
    var chosenWord = selecteddictionary[randomNumber].name
    for (var i = 0; i < chosenWord.length; i++) {
        arrayOfChosenWord[i] = chosenWord.charAt(i).toUpperCase()
    }
    for (var e = 0; e < arrayOfChosenWord.length; e++) {
        if (arrayOfChosenWord[e] == " " || arrayOfChosenWord[e] == "-" ) {
            wordLinesArray[e] = arrayOfChosenWord[e]
        }
        else {
            wordLinesArray[e] = "_"
        }
    }
    wordLines = wordLinesArray.join(" ")
    console.log("\n", wordLines)
    return randomNumber
}

// to get a hint if necessary
function getHint(randomNumber) {
    var hintOfChosenWord = selecteddictionary[randomNumber].hint
    var needHint = readline.question(hintPrompt)
    if (needHint.includes("y")) {
        console.log(hintOfChosenWord)
    }
    else if (needHint.includes("n")) {
    }
    else {
        errorMsg2()
        getHint(randomNumber)
    }

}

// game logic
function guessWord() {
    guess = readline.question(guessPrompt)
    if (arrayOfChosenWord.join('').includes(guess.toUpperCase())) {
        if (guess.length == 0) {
            errorMsg1()
        }
        else if (guess.length > 1 && guess.split('').length == arrayOfChosenWord.length) {
            if (guess.toUpperCase() == arrayOfChosenWord.join('')) {
                won = 1
                wonMsg()
                replay()
            }
        }
        else if (guess.length > 1 && guess.split('').length != arrayOfChosenWord.length) {
            tries -= 1
            triesLeftMsg(tries)
            errorMsg1()
        }
        else if (guess.length = 1) {
            for (var i = 0; i < arrayOfChosenWord.length; i++){
                if (arrayOfChosenWord[i].includes(guess.toUpperCase())) {
                    wordLinesArray[i] = arrayOfChosenWord[i]
                    if (wordLinesArray.includes("_")){
                        console.log("\n", wordLinesArray.join(" "))
                    }
                    else if (won == 0) {
                        won = 1
                        console.log("\n", wordLinesArray.join(" "))
                        wonMsg()
                        replay()
                    }
                    else {
                        console.log("error 1")
                    }
                }
            }
        }
        else {
            errorMsg1()
        }
    }
    else if (regex.test(guess) == true) {
        tries -= 1
        triesLeftMsg(tries)
    }
    else {
        errorMsg1()
    }
}

function main() {
    reset()
    pickCat()
    randomNumber = randomWordPicker()
    while (tries > 0 && won == 0) {
        if (tries == 3 && hints > 0) {
            getHint(randomNumber)
            hints -= 1
        }
        guessWord()
    }
    if (tries == 0) {
        lostMsg()
        replay()
    }
}

menu()
